contracts/neutron-interchain-txs/src/contract.rs

```diff
@@ -0,0 +1,474 @@
+        },
+    )?;
+
+    Ok(Response::default().add_submessages(vec![submsg]))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn sudo(deps: DepsMut, env: Env, msg: SudoMsg) -> StdResult<Response> {
+    match msg {
+        SudoMsg::Response { request, data } => sudo_response(deps, request, data),
+        SudoMsg::Error { request, details } => sudo_error(deps.as_ref(), request, details),
+        SudoMsg::Timeout { request } => sudo_timeout(deps, env, request),
+        SudoMsg::OpenAck {
+            port_id,
+            channel_id,
+            counterparty_channel_id,
+            counterparty_version,
+        } => sudo_open_ack(
+            deps,
+            env,
+            port_id,
+            channel_id,
+            counterparty_channel_id,
+            counterparty_version,
+        ),
+        _ => Ok(Response::default()),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
+    Ok(Response::default())
+}
+
+fn sudo_open_ack(
+    deps: DepsMut,
+    _env: Env,
+    port_id: String,
+    _channel_id: String,
+    _counterparty_channel_id: String,
+    counterparty_version: String,
+) -> StdResult<Response> {
+    let parsed_version: Result<OpenAckVersion, _> =
+        serde_json_wasm::from_str(counterparty_version.as_str());
+    if let Ok(parsed_version) = parsed_version {
+        INTERCHAIN_ACCOUNTS.save(
+            deps.storage,
+            port_id,
+            &Some((
+                parsed_version.address,
+                parsed_version.controller_connection_id,
+            )),
+        )?;
+        return Ok(Response::default());
+    }
+    Err(StdError::generic_err("Can't parse counterparty_version"))
+}
+
+fn sudo_response(deps: DepsMut, request: RequestPacket, data: Binary) -> StdResult<Response> {
+    deps.api.debug(
+        format!(
+            "WASMDEBUG: sudo_response: sudo received: {:?} {:?}",
+            request, data
+        )
+        .as_str(),
+    );
+    let seq_id = request
+        .sequence
+        .ok_or_else(|| StdError::generic_err("sequence not found"))?;
+
+    let payload = read_sudo_payload(deps.storage, seq_id)?;
+    deps.api
+        .debug(format!("WASMDEBUG: sudo_response: sudo payload: {:?}", payload).as_str());
+    // handle response
+    let parsed_data = parse_response(data)?;
+
+    for item in parsed_data {
+        let item_type = item.msg_type.as_str();
+        match item_type {
+            "/cosmos.staking.v1beta1.MsgUndelegate" => {
+                let out: MsgUndelegateResponse = parse_item(&item.data)?;
+                let completion_time = out
+                    .completion_time
+                    .ok_or_else(|| StdError::generic_err("failed to get completion time"))?;
+                deps.api
+                    .debug(format!("Undelegation completion time: {:?}", completion_time).as_str());
+            }
+            _ => {
```

Comment:
not obvious, our `Delegate` response comes here. I would suggest either add a comment or add a branch with `/cosmos.staking.v1beta1.MsgDelegate` item_type

Reply:
well, I added it in my task, so probably unnecessary